src/main/groovy/org/web3j/solidity/gradle/plugin/ImportsResolver.groovy

```diff
@@ -12,21 +12,31 @@
  */
 package org.web3j.solidity.gradle.plugin
 
-import groovy.transform.Memoized
+import groovy.io.FileType
+
+import java.util.regex.Pattern
 
 /**
  * Helper class to resolve the external imports from a Solidity file.
  *
- * Supported providers are:
+ * The import resolving is done in three steps:
  * <ul>
- *     <li><a href="https://www.npmjs.com/package/@openzeppelin/contracts">Open Zeppelin</a></li>
- *     <li><a href="https://www.npmjs.com/package/@uniswap/lib">Uniswap</a></li>
+ *     <li>
+ *         First, all packages needed for direct imports are extracted from sol files to generate a package.json.
+ *         This is done in a separate Gradle task, so that the next steps are only performed when direct imports change.
+ *     </li>
+ *     <li>
+ *         Second, required packages are downloaded by npm.
+ *     </li>
+ *     <li>
+ *         Third, sol files that were downloaded are analyzed as well and all packages required are collected.
+ *         This information is used in compileSolidity for the allowed paths and the path remappings.
+ *     </li>
  * </ul>
  */
-@Singleton
```

Comment:
`@Singleton` and `@Memoized` are problematic in Gradle plugins. Plugin classes are loaded in the Gradle daemon process and are then reused between build runs. Static state is then preserved. I observed here that `extractImports` still returned old values, even though I changed the source file and the task did run again.